### PR TITLE
dev: Maint & CSS/JS resource caching

### PR DESCRIPTION
**Pull request by kingthorin**
## Overview
- CHANGELOG > Added note.
- PerformacneDir > Correct extended class.
- SequencePage > Add constants and extract methods to simplify code.
- TestDirectory > Changes to facilitate caching.
- TestProxyServer > Changes to facilitate caching.

## Related Issues
n/a

## Checklist
- [na] Update help
- [x] Update changelog
- [x] Run `./gradlew spotlessApply` for code formatting
- [na] Write tests
- [na] Check code coverage
- [x] Sign-off commits
- [x] Squash commits
- [x] Use a descriptive title

### CODE REVIEW

File: addOns/dev/src/main/java/org/zaproxy/addon/dev/seq/simple/SimpleSequencePage.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.network.HttpMalformedHeaderException;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.network.HttpRequestHeader;
+import org.zaproxy.addon.dev.TestPage;
+import org.zaproxy.addon.dev.TestProxyServer;
+import org.zaproxy.addon.network.server.HttpMessageHandlerContext;
+
+public class SimpleSequencePage extends TestPage {
```
[the remark this answers is not in the document]
Comment (kingthorin) [reply addOns/dev/src/main/java/org/zaproxy/addon/dev/seq/simple/SimpleSequencePage.java:35]: Agreed, Simon and I discussed that yesterday 🙂

File: addOns/dev/src/main/java/org/zaproxy/addon/dev/TestProxyServer.java [hunk 2]
```diff
@@ -165,17 +171,37 @@ public String getTextFile(TestDirectory dir, String name) {
...
-        sb.append("Pragma: no-cache\r\n");
-        sb.append("Cache-Control: no-cache, no-store, must-revalidate\r\n");
+        if (allowCache) {
+            sb.append("Cache-Control: public\r\n");
+            sb.append(
+                    "Expires: "
+                            + DateTimeFormatter.RFC_1123_DATE_TIME
+                                    .withZone(ZoneOffset.UTC)
+                                    .format(Instant.now().plus(7, ChronoUnit.DAYS))
+                            + "\r\n");
```
Comment (kingthorin) [inline addOns/dev/src/main/java/org/zaproxy/addon/dev/TestProxyServer.java:200]: Should use append vs concatenation

File: addOns/dev/src/main/java/org/zaproxy/addon/dev/TestProxyServer.java [hunk 3]
```diff
@@ -165,25 +171,45 @@ public String getTextFile(TestDirectory dir, String name) {
...
+            sb.append("Cache-Control: no-cache, no-store, must-revalidate").append(HttpHeader.CRLF);
+        }
+        sb.append("Access-Control-Allow-Methods: GET,POST,OPTIONS").append(HttpHeader.CRLF);
+        sb.append("Access-Control-Allow-Headers: ZAP-Header").append(HttpHeader.CRLF);
+        sb.append("X-Frame-Options: DENY").append(HttpHeader.CRLF);
+        sb.append("X-XSS-Protection: 1; mode=block").append(HttpHeader.CRLF);
+        sb.append("X-Content-Type-Options: nosniff").append(HttpHeader.CRLF);
+        sb.append("X-Clacks-Overhead: GNU Terry Pratchett").append(HttpHeader.CRLF);
+        sb.append("Content-Length: ").append(contentLength).append(HttpHeader.CRLF);
+        sb.append("Content-Type: ").append(contentType).append(HttpHeader.CRLF);
```
Comment (kingthorin) [inline addOns/dev/src/main/java/org/zaproxy/addon/dev/TestProxyServer.java:212]: Leverage `HttpHeader.CRLF`

**Comment by kingthorin** [comment]
image

File: addOns/dev/src/main/java/org/zaproxy/addon/dev/TestProxyServer.java [hunk 4]
```diff
@@ -165,25 +171,45 @@ public String getTextFile(TestDirectory dir, String name) {
...
+                    .append(
+                            DateTimeFormatter.RFC_1123_DATE_TIME
+                                    .withZone(ZoneOffset.UTC)
+                                    .format(Instant.now().plus(7, ChronoUnit.DAYS)))
+                    .append(HttpHeader.CRLF);
+        } else {
+            sb.append("Pragma: no-cache").append(HttpHeader.CRLF);
+            sb.append("Cache-Control: no-cache, no-store, must-revalidate").append(HttpHeader.CRLF);
+        }
+        sb.append("Access-Control-Allow-Methods: GET,POST,OPTIONS").append(HttpHeader.CRLF);
```
Comment (thc202) [inline addOns/dev/src/main/java/org/zaproxy/addon/dev/TestProxyServer.java:205]: No advantage in splitting the existing strings just to use the constant.

File: addOns/dev/CHANGELOG.md [hunk 5]
```diff
@@ -6,6 +6,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ## Unreleased
 ### Changed
 - Sequence performance test to make it actually possible to test it using automation.
+- css and js responses are now set cache enabled.
```
Comment (thc202) [inline addOns/dev/CHANGELOG.md:9]: CSS and JS

File: addOns/dev/src/main/java/org/zaproxy/addon/dev/TestProxyServer.java [hunk 6]
```diff
@@ -165,17 +169,38 @@ public String getTextFile(TestDirectory dir, String name) {
...
+    public static String getDefaultResponseHeader(
+            String responseStatus, String contentType, int contentLength, boolean allowCache) {
         StringBuilder sb = new StringBuilder(250);
 
         sb.append("HTTP/1.1 ").append(responseStatus).append("\r\n");
-        sb.append("Pragma: no-cache\r\n");
-        sb.append("Cache-Control: no-cache, no-store, must-revalidate\r\n");
+
+        if (allowCache) {
+            sb.append("Cache-Control: public\r\n").append(HttpHeader.CRLF);
```
Comment (thc202) [inline addOns/dev/src/main/java/org/zaproxy/addon/dev/TestProxyServer.java:193]: Adding too many CRLF.

File: addOns/dev/src/main/java/org/zaproxy/addon/dev/seq/performance/SequencePage.java [hunk 7]
```diff
@@ -121,81 +123,15 @@ public void handleMessage(HttpMessageHandlerContext ctx, HttpMessage msg) {
 
             if (checkSequence && expectedStep == null) {
                 // Unknown sequence
```
Comment (thc202) [inline addOns/dev/src/main/java/org/zaproxy/addon/dev/seq/performance/SequencePage.java:125]: These comments are now redundant with the methods being called.

File: addOns/dev/src/main/java/org/zaproxy/addon/dev/seq/performance/SequencePage.java [hunk 8]
```diff
@@ -208,6 +144,91 @@ public void handleMessage(HttpMessageHandlerContext ctx, HttpMessage msg) {
         }
     }
 
+    void handleIntermediateStep(
```
Comment (thc202) [inline addOns/dev/src/main/java/org/zaproxy/addon/dev/seq/performance/SequencePage.java:147]: I'd keep them private for the time being.

File: addOns/dev/src/main/java/org/zaproxy/addon/dev/seq/performance/SequencePage.java [hunk 9]
```diff
@@ -208,6 +144,91 @@ public void handleMessage(HttpMessageHandlerContext ctx, HttpMessage msg) {
...
+        sb.append("Out of sequence step!\n");
+        sb.append("Got ");
+        sb.append(seqStep);
+        sb.append(" but expected ");
+        sb.append(seqMap.get(seqUuid));
+        sb.append(PARA);
+        sb.append("Sorry, but you have to <a href=\"seq\">Start Again</a>");
+        sb.append(PARA);
+
+        responseStatus = TestProxyServer.STATUS_FORBIDDEN;
```
Comment (thc202) [inline addOns/dev/src/main/java/org/zaproxy/addon/dev/seq/performance/SequencePage.java:217]: Just return it.

File: addOns/dev/src/main/java/org/zaproxy/addon/dev/seq/performance/SequencePage.java [hunk 10]
```diff
@@ -208,6 +144,91 @@ public void handleMessage(HttpMessageHandlerContext ctx, HttpMessage msg) {
...
+    }
+
+    String handledUnknownSequence(StringBuilder sb) {
+        String responseStatus;
+        sb.append("Unregistered sequence!\n");
+        sb.append(PARA);
+        sb.append("Sorry, but you have to <a href=\"seq\">Start Again</a>");
+        sb.append(PARA);
+
+        responseStatus = TestProxyServer.STATUS_FORBIDDEN;
```
Comment (thc202) [inline addOns/dev/src/main/java/org/zaproxy/addon/dev/seq/performance/SequencePage.java:228]: Same here.

**Comment by kingthorin** [comment]
Fixed

**Review by thc202 — APPROVED** [review]

**Review by psiinon — APPROVED** [review]

**Pull request merged**